services/oauth.js

```diff
@@ -0,0 +1,166 @@
+var request = require('superagent');
+var crypto = require('crypto');
+
+/**
+ * Returns a nonce for the timestamp `time`. OAuth 1.0 defines a
+ * nonce as a value unique to a given timestamp.
+ * @param {Number} time
+ * @returns {String}
+ * @see https://oauth.net/core/1.0a/#nonce
+ * @private
+ */
+
+function nonce(time) {
+	return crypto.createHash('sha1').update(String(time)).digest('hex');
+}
+
+/**
+ * Returns the HMAC-SHA1 digest of `text` and `key` in baset64 encoding.
+ * @param {String} text
+ * @param {String} key
+ * @returns {String}
+ * @see https://oauth.net/core/1.0a/#rfc.section.9.2
+ * @private
+ */
+
+function hmac(text, key) {
+	return crypto.createHmac('sha1', key).update(text).digest('base64');
+}
+
+/**
+ * Encodes each of the strings in `arr` and joins them with the '&'
+ * character (ASCII code 38).
+ * @param {String[]} arr
+ * @returns {String}
+ * @see https://oauth.net/core/1.0a/#rfc.section.9.1.3
+ * @see https://oauth.net/core/1.0a/#encoding_parameters
+ * @private
+ */
+
+function join(arr) {
+	return arr.map(encodeURIComponent).join('&');
+}
+
+/**
+ * Normalizes and sorts all of the query string params in `req`,
+ * then joins them with the '&' character (ASCII code 38).
+ * @param {Request}
+ * @returns String.
+ * @see https://oauth.net/core/1.0a/#sig_norm_param
+ * @private
+ */
+
+function query(req) {
+	var pairs = req.qsRaw.slice();
+
+	Object.keys(req.qs).forEach(function (key) {
+		pairs.push(key + '=' + encodeURIComponent(req.qs[key]));
```

Comment:
does req.qs already normalize objects to strings/numeric, or do we need to JSON.stringify objects here too?